File: Bakcend/controllers/productController.js
const Product = require('../models/productModel');
const cloudinary = require('cloudinary').v2;
const User = require("../models/userModel");

cloudinary.config({
  cloud_name: 'dhqqt1nju',
  api_key: '825414438411693',
  api_secret: 'Xz3HC1YaOrsSYHaBpGyFkn7hbZQ',
});
const path = require('path');
const createProduct = async (req, res) => {
  try {
    const product = new Product({
      sellerId: req.body.sellerId,
      nomProduit: req.body.nomProduit,
      prixProduitAfter: req.body.prixProduitAfter,
      prixProduitBefore: req.body.prixProduitBefore,
      photoProduitPrincipale: '',
      photosSecondaire: [],
      description: req.body.description,
    });

    let result = '';

    if (req.file && req.file.path) {
      result = await cloudinary.uploader.upload(req.file.path);
    }

    product.photoProduitPrincipale = result.secure_url || '';

    if (req.files && req.files.length > 0) {
      for (const file of req.files) {
        const photo = await cloudinary.uploader.upload(file.path);
        product.photosSecondaire.push(photo.secure_url);
      }
    }

    await product.save();

    // Update the user's myProducts array with the newly created product's ID
    const user = await User.findById(req.body.sellerId);
    if (user) {
      user.myProducts.push(product._id);
      await user.save();
    }

    res.status(200).json({
      sellerId: product.sellerId,
      id: product._id,
      message: 'Product created successfully',
    });
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error creating product' });
  }
};




const getProductById = async (req, res) => {
  try {
    const productId = req.params.id;
    const product = await Product.findById(productId);
    if (!product) {
      return res.status(404).json({ message: 'Product not found' });
    }
    res.status(200).json(product);
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error fetching product' });
  }
};


//delete product
const deleteAllProducts = async (req, res) => {
  try {
    await Product.deleteMany({});
    res.status(200).json({ message: 'All products deleted successfully' });
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error deleting products' });
  }
};


//get all products
const getAllProducts = async (req, res) => {
  try {
    const products = await Product.find({});
    res.status(200).json(products);
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error fetching products' });
  }
};



const deleteProductById = async (req, res) => {
  try {
    const productId = req.params.id;

    // Find the product by ID
    const product = await Product.findById(productId);

    if (!product) {
      return res.status(404).json({ message: 'Product not found' });
    }

    // Delete the product from the database
    await Product.deleteOne({ _id: productId });

    // Remove the product ID from the user's myProducts array
    const user = await User.findById(product.sellerId);
    if (user) {
      user.myProducts = user.myProducts.filter(id => id.toString() !== productId);
      await user.save();
    }

    res.status(200).json({ message: 'Product deleted successfully' });
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error deleting product' });
  }
};


const updateProductById = async (req, res) => {
  try {
    const productId = req.params.id;

    // Find the product by ID
    const product = await Product.findById(productId);

    if (!product) {
      return res.status(404).json({ message: 'Product not found' });
    }

    // Update the product's fields with new values from the request body
    product.nomProduit = req.body.nomProduit || product.nomProduit;
    product.prixProduitAfter = req.body.prixProduitAfter || product.prixProduitAfter;
    product.prixProduitBefore = req.body.prixProduitBefore || product.prixProduitBefore;
    product.description = req.body.description || product.description;

    // Save the updated product
    await product.save();

    res.status(200).json({ message: 'Product updated successfully', updatedProduct: product });
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: 'Error updating product' });
  }
};



module.exports = { createProduct,getProductById,deleteAllProducts,getAllProducts,deleteProductById,updateProductById };

File: bioshop/lib/controllers/productController.dart
import 'dart:convert';
import 'dart:io';
import 'package:bioshop/models/Product.dart';
import 'package:bioshop/views/Home.dart';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:http/http.dart' as http;


import '../api/Api.dart';

class ProductController extends GetxController {
  Product? productData;
  RxList<Product> products = <Product>[].obs;

  @override
  void onInit() {
    super.onInit();
    fetchProducts();

  }

  Future<void> fetchProducts() async {
    try {
      final List<Product> fetchedProducts = await getAllProducts();
      products.assignAll(fetchedProducts);
    } catch (e) {
      print('Error fetching products: $e');
    }
  }

  Future<List<Product>> getAllProducts() async {
    final url = '${API().url.toString()}product';

    try {
      final response = await http.get(Uri.parse(url));

      if (response.statusCode == 200) {
        final List<dynamic> jsonResponse = json.decode(response.body);
        return jsonResponse.map((item) => Product.fromJson(item)).toList();
      } else {
        throw Exception('Failed to load products');
      }
    } catch (e) {
      throw Exception('Error fetching products: $e');
    }
  }



Future<Product> getProductById(String productId) async {
  final String apiUrl = '${API().url.toString()}product/$productId'; // Replace with your actual API endpoint

  try {
    final response = await http.get(Uri.parse(apiUrl));

    if (response.statusCode == 200) {
      final Map<String, dynamic> productData = json.decode(response.body);
      return Product.fromJson(productData);
    } else if (response.statusCode == 404) {
      throw Exception('Product not found');
    } else {
      throw Exception('Error fetching product');
    }
  } catch (error) {
    throw Exception('Error: $error');
  }
}



  
Future<void> sellProduct(
  String sellerId,
 String nomProduit,
 String prixProduitAfter,
String prixProduitBefore,
String description,
File imageFile) async {
  final url = Uri.parse('${API().url.toString()}product/addProduct');

  // Create a multipart request for the image upload
  var request = http.MultipartRequest('POST', url);

  if (imageFile != null) {
    // Attach the image file to the request
    request.files.add(await http.MultipartFile.fromPath('photoProduitPrincipale', imageFile.path));
  } else {
    print('(null)');
  }

  // Add the fields to the request body
  request.fields['sellerId'] = sellerId;
  request.fields['nomProduit'] = nomProduit;
  request.fields['prixProduitAfter'] = prixProduitAfter;
   request.fields['prixProduitBefore'] = prixProduitBefore;
    request.fields['description'] = description;


  try {
    final response = await request.send();

    // Handle different response statuses
    if (response.statusCode == 200) {
      // Registration successful
      print('Registration successful');
      Get.snackbar(
        "Success",
        "Product added successfully",
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor:  Colors.green,
        colorText: Colors.white,
      );
      await fetchProducts();
      Get.to(HomePage());
    } else if (response.statusCode == 400) {
      // Bad request
      print('Bad request');
 
      
    } else {
     
    }
  } catch (e) {
  }
}



Future<void> deleteProductById(String productId) async {
    final String apiUrl = '${API().url.toString()}product/id/$productId'; // Replace with your actual API endpoint

    try {
      final response = await http.delete(Uri.parse(apiUrl));

      if (response.statusCode == 200) {
        Get.snackbar(
        "Success",
        "Product Deleted Successfully",
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor:  Colors.green,
        colorText: Colors.white,
      );
      fetchProducts();
       Get.to(HomePage());
        print('Product deleted successfully');
      } else if (response.statusCode == 404) {
        throw Exception('Product not found');
      } else {
        throw Exception('Error deleting product');
      }
    } catch (error) {
      throw Exception('Error: $error');
    }
  }


  Future<void> updateProductById(
    String productId,
    String nomProduit,
    String prixProduitAfter,
    String prixProduitBefore,
    String description,
  ) async {
    final String apiUrl = '${API().url.toString()}product/$productId'; // Replace with your actual API endpoint

    final updatedData = {
      'nomProduit': nomProduit,
      'prixProduitAfter': prixProduitAfter,
      'prixProduitBefore': prixProduitBefore,
      'description': description,
    };

    try {
      final response = await http.put(
        Uri.parse(apiUrl),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(updatedData),
      );

      if (response.statusCode == 200) {
        Get.snackbar(
        "Success",
        "Product updated successfully",
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor:  Colors.green,
        colorText: Colors.white,
      );
        print('Product updated successfully');
        Get.to(HomePage());
      } else if (response.statusCode == 404) {
       
        throw Exception('Product not found');
      } else {
          Get.snackbar(
        "Error",
        "Error updating product",
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor:  Colors.red,
        colorText: Colors.white,
      );
        throw Exception('Error updating product');
      }
    } catch (error) {
      throw Exception('Error: $error');
    }
  }



}
